### PR TITLE
Configure remote wrote agent ingress

### PR DESCRIPTION
**Pull request by QuentinBisson**
Needed so the remote write agent can send data to prometheus securely

## Checklist

I have:

- [x] Described why this change is being introduced
- [x] Separated out refactoring/reformatting in a dedicated PR
- [x] Updated changelog in `CHANGELOG.md`

### CODE REVIEW

**Comment by QuentinBisson** [comment]
Tested on Gauss

File: CHANGELOG.md
Comment (mcbenjemaa) [inline CHANGELOG.md:15]: Those must be rebased after merging #1000

Comment (mcbenjemaa) [reply CHANGELOG.md:15]: Please consider merging #1000  first.

Comment (mcbenjemaa) [reply CHANGELOG.md:15]: it makes me a little bit confused, as it already includes changes of #1000

File: service/controller/resource/monitoring/remotewriteingress/v1/resource.go
Comment (mcbenjemaa) [inline service/controller/resource/monitoring/remotewriteingress/v1/resource.go:65]: Can you have a function that generates those annotation for basic auth

File: service/controller/resource/monitoring/remotewriteingress/v1/resource.go
Comment (mcbenjemaa) [inline service/controller/resource/monitoring/remotewriteingress/v1/resource.go:122]: Can you put the name of service as constant.

File: service/controller/resource/monitoring/remotewriteingress/v1beta1/resource.go
Comment (mcbenjemaa) [inline service/controller/resource/monitoring/remotewriteingress/v1beta1/resource.go:66]: Same, please put those annotation in a function.

File: service/controller/resource/monitoring/remotewriteingress/v1beta1/resource.go
Comment (mcbenjemaa) [inline service/controller/resource/monitoring/remotewriteingress/v1beta1/resource.go:119]: Same remark for name of the service!

File: CHANGELOG.md
Comment (QuentinBisson) [reply CHANGELOG.md:15]: Because it's not possible to have them separate and this saves me time to wait for reviews 😁

File: service/controller/resource/monitoring/remotewriteingress/v1/resource.go
Comment (QuentinBisson) [reply service/controller/resource/monitoring/remotewriteingress/v1/resource.go:65]: Can you explain why? Considering I'm planning on removing v1beta1, I'd rather not :)

File: service/controller/resource/monitoring/remotewriteingress/v1/resource.go
Comment (QuentinBisson) [reply service/controller/resource/monitoring/remotewriteingress/v1/resource.go:122]: Sure

File: service/controller/resource/monitoring/remotewriteingress/v1/resource.go
Comment (mcbenjemaa) [reply service/controller/resource/monitoring/remotewriteingress/v1/resource.go:65]: I don’t like hardcoded things like this!

Comment (QuentinBisson) [reply service/controller/resource/monitoring/remotewriteingress/v1/resource.go:65]: TBH, hardcoding it here or in a function is the same to me :D

Comment (QuentinBisson) [reply service/controller/resource/monitoring/remotewriteingress/v1/resource.go:65]: Done

File: service/controller/resource/monitoring/remotewriteagentconfigsecret/resource.go
Comment (mcbenjemaa) [inline service/controller/resource/monitoring/remotewriteagentconfigsecret/resource.go:60]: Don't we need to add some labels?

File: service/controller/resource/monitoring/remotewritesecret/resource.go
Comment (mcbenjemaa) [inline service/controller/resource/monitoring/remotewritesecret/resource.go:1]: I'm a bit confused 😕 
Why this package? `remotewritesecret` while the logic is added in `remotewriteagentsecret`

Comment (QuentinBisson) [reply service/controller/resource/monitoring/remotewritesecret/resource.go:1]: I think this one needs to go 😓 
Too many refactoring between the first and the last iteration

File: service/controller/resource/monitoring/remotewriteagentconfigsecret/doc.go
Comment (TheoBrigitte) [inline service/controller/resource/monitoring/remotewriteagentconfigsecret/doc.go:4]: If I got this correctly in this package we are wrapping the `RemoteWrite` config in a secret which will then be used by prometheus-operator to configure the agent. Is this right ?

File: service/controller/resource/monitoring/remotewriteagentconfigsecret/resource.go
Comment (mcbenjemaa) [reply service/controller/resource/monitoring/remotewriteagentconfigsecret/resource.go:60]: Do we really need to add labels like. 
```
clusterID
organization
Installation 
etc? 
```

**Comment by mcbenjemaa** [comment]
@QuentinBisson some changes are included here https://github.com/giantswarm/prometheus-meta-operator/pull/1000
?

File: service/controller/resource/monitoring/remotewriteagentconfigsecret/resource.go
Comment (QuentinBisson) [reply service/controller/resource/monitoring/remotewriteagentconfigsecret/resource.go:60]: I don't see which other labels to set tbh

Comment (QuentinBisson) [reply service/controller/resource/monitoring/remotewriteagentconfigsecret/resource.go:60]: So i'd rather have suggestions :)

**Comment by QuentinBisson** [comment]
Yes I already explained that in the previous PRs :D

**Comment by QuentinBisson** [comment]
#1001 is reliant on #1000 so #1000 needs to be merged first

**Comment by mcbenjemaa** [comment]
I will move my reviews notes to #1000

File: service/controller/resource/monitoring/remotewriteagentconfigsecret/resource.go
Comment (TheoBrigitte) [inline service/controller/resource/monitoring/remotewriteagentconfigsecret/resource.go:78]: Why is this an array?

File: service/controller/resource/monitoring/remotewriteagentconfigsecret/test/case-1-awsconfig.golden
Comment (TheoBrigitte) [inline service/controller/resource/monitoring/remotewriteagentconfigsecret/test/case-1-awsconfig.golden:2]: We do not really see anything here. This does not bring much value.

File: pkg/password/manager.go
Comment (TheoBrigitte) [inline pkg/password/manager.go:28]: Why are we hashing the password through bcrypt here ?

File: service/controller/resource/monitoring/remotewriteagentsecret/resource.go
Comment (TheoBrigitte) [inline service/controller/resource/monitoring/remotewriteagentsecret/resource.go:91]: Can we leave this magic to the client/prometheus-agent ? I would rather that we provider username and password and the client use them as he wants to.

Also why are we sending a hashed password here?

File: service/controller/resource/monitoring/remotewriteagentsecret/test/case-1-awsconfig.golden
Comment (TheoBrigitte) [inline service/controller/resource/monitoring/remotewriteagentsecret/test/case-1-awsconfig.golden:4]: again we do not see much what is being produced here, this does not bring much value.

File: service/key/key.go
Comment (TheoBrigitte) [inline service/key/key.go:22]: In general I would prefer to avoid being too specific. We are just configuring a remote write endpoint here. We do not need to mention the `agent` thing here. It could be consumed by a standard Prometheus also, and this might changes. Let's avoid being too specific please. This is also valid for package names.

**Review by TheoBrigitte — CHANGES_REQUESTED** [review]
Good job overall :+1:

File: service/controller/resource/monitoring/remotewriteagentconfigsecret/doc.go
Comment (QuentinBisson) [reply service/controller/resource/monitoring/remotewriteagentconfigsecret/doc.go:4]: We could eventually fill this with the remotewrite CR as well

File: service/controller/resource/monitoring/remotewriteagentconfigsecret/resource.go
Comment (QuentinBisson) [reply service/controller/resource/monitoring/remotewriteagentconfigsecret/resource.go:78]: Because the prometheus agent app expects an array as we can use it to send remote write data to customers in the end

File: service/controller/resource/monitoring/remotewriteagentconfigsecret/test/case-1-awsconfig.golden
Comment (QuentinBisson) [reply service/controller/resource/monitoring/remotewriteagentconfigsecret/test/case-1-awsconfig.golden:2]: What would you like to have here? No tests?

File: pkg/password/manager.go
Comment (QuentinBisson) [reply pkg/password/manager.go:28]: 1: Because it needs to be hashed
2: because nginx does not support a lot of hashes
3: because bcrypt is one of the best security hashes out there

File: service/controller/resource/monitoring/remotewriteagentsecret/resource.go
Comment (QuentinBisson) [reply service/controller/resource/monitoring/remotewriteagentsecret/resource.go:91]: No we cannot because this auth parameter is used by the ingress on the management cluster and the password needs to be hashed for nginx to work

File: service/controller/resource/monitoring/remotewriteagentsecret/test/case-1-awsconfig.golden
Comment (QuentinBisson) [reply service/controller/resource/monitoring/remotewriteagentsecret/test/case-1-awsconfig.golden:4]: I'd rather have minimal tests than no tests at all

File: service/key/key.go
Comment (QuentinBisson) [reply service/key/key.go:22]: Except we are not just configuring a remote write endpoint, we are configuring it for the agent. We already have remotewritesecret and remotewriteconfig as resources anyway so we need to avoid collision between the 2

File: service/controller/resource/monitoring/remotewriteagentconfigsecret/resource.go
Comment (QuentinBisson) [reply service/controller/resource/monitoring/remotewriteagentconfigsecret/resource.go:78]: I would rather we hare too broad here instead of too narrow

**Comment by QuentinBisson** [comment]
@TheoBrigitte can you check my answer to your comments?

File: service/controller/resource/monitoring/remotewriteagentconfigsecret/resource.go
Comment (TheoBrigitte) [reply service/controller/resource/monitoring/remotewriteagentconfigsecret/resource.go:78]: Ok good.

File: service/controller/resource/monitoring/remotewriteagentconfigsecret/test/case-1-awsconfig.golden
Comment (TheoBrigitte) [reply service/controller/resource/monitoring/remotewriteagentconfigsecret/test/case-1-awsconfig.golden:2]: Having the content of the values would be enough I guess.

File: pkg/password/manager.go
Comment (TheoBrigitte) [reply pkg/password/manager.go:28]: I though basic auth where just base64 encoded, maybe I am wrong.

File: service/key/key.go
Comment (TheoBrigitte) [reply service/key/key.go:22]: Maybe something like `incoming-remote-write` ? `agent` feels wrong here, this is just a mode used to run Prometheus.

File: service/controller/resource/monitoring/remotewriteagentsecret/test/case-1-awsconfig.golden
Comment (TheoBrigitte) [reply service/controller/resource/monitoring/remotewriteagentsecret/test/case-1-awsconfig.golden:4]: But the test should produce output that we can compare against input in code.

File: pkg/password/manager.go
Comment (QuentinBisson) [reply pkg/password/manager.go:28]: Sadly no: https://docs.giantswarm.io/advanced/ingress/configuration/#authentication It requires a hash of the password. In the simplest example, it uses apache `htpasswd` but there is no go implementation to cypher the password so I picked the best in go that nginx supports

File: service/controller/resource/monitoring/remotewriteagentconfigsecret/test/case-1-awsconfig.golden
Comment (QuentinBisson) [reply service/controller/resource/monitoring/remotewriteagentconfigsecret/test/case-1-awsconfig.golden:2]: That's encoded in base 64 by the secret creation mechanism. The code will not allow us to generate non base64 secret just for the tests

**Comment by QuentinBisson** [comment]
@TheoBrigitte apart from the secret encoded in base64, the rest should be addressed :)

File: service/controller/resource/monitoring/remotewriteagentconfigsecret/test/case-1-awsconfig.golden
Comment (TheoBrigitte) [reply service/controller/resource/monitoring/remotewriteagentconfigsecret/test/case-1-awsconfig.golden]: Maybe we can have a totally different test that just test the part that generate the values before the secret then.

File: pkg/password/manager.go
Comment (TheoBrigitte) [reply pkg/password/manager.go:28]: Ok fine

**Review by TheoBrigitte — APPROVED** [review]
I would like to see better tests output but I won't block this longer.

**Comment by QuentinBisson** [comment]
Let's improve the tests once we have a working feature :)

**Pull request merged**
